src/shared/codelens/csharpCodeLensProvider.ts

```diff
@@ -0,0 +1,305 @@
+    document: vscode.TextDocument,
+    symbols: vscode.DocumentSymbol[],
+): DotNetHandlerSymbolsTuplet[] {
+    return symbols
+        .filter(symbol => symbol.kind === vscode.SymbolKind.Namespace)
+        // Find relevant classes within the namespace
+        .reduce<{
+            namespace: vscode.DocumentSymbol,
+            class: vscode.DocumentSymbol,
+        }[]>(
+            (accumulator, namespaceSymbol: vscode.DocumentSymbol) => {
+                accumulator.push(...namespaceSymbol.children
+                    .filter(namespaceChildSymbol => namespaceChildSymbol.kind === vscode.SymbolKind.Class)
+                    .filter(methodSymbol => isPublicClassSymbol(document, methodSymbol))
+                    .map(classSymbol => {
+                        return {
+                            namespace: namespaceSymbol,
+                            class: classSymbol,
+                        }
+                    })
+                )
+
+                return accumulator
+            },
+            []
+        )
+        // Find relevant methods within each class
+        .reduce<DotNetHandlerSymbolsTuplet[]>(
+            (accumulator, tuplet) => {
+                accumulator.push(...tuplet.class.children
+                    .filter(classChildSymbol => classChildSymbol.kind === vscode.SymbolKind.Method)
+                    .filter(methodSymbol => getMethodNameFromSymbol(methodSymbol) !== undefined)
+                    .filter(methodSymbol => isPublicMethodSymbol(document, methodSymbol))
+                    .map(methodSymbol => {
+                        return {
+                            namespace: tuplet.namespace,
+                            class: tuplet.class,
+                            method: methodSymbol,
+                        }
+                    })
+                )
+
+                return accumulator
+            },
+            []
+        )
+}
+
+export async function findParentProjectFile(
+    sourceCodeUri: vscode.Uri,
+    findWorkspaceFiles: typeof vscode.workspace.findFiles = vscode.workspace.findFiles,
+): Promise<vscode.Uri | undefined> {
+    const workspaceProjectFiles: vscode.Uri[] = await findWorkspaceFiles(
+        '**/*.csproj'
+    )
+
+    // Use the project file "closest" in the parent chain to sourceCodeUri
+    // Assumption: only one .csproj file will exist in a given folder
+    const parentProjectFiles = workspaceProjectFiles
+        .filter(uri => {
+            const dirname = dirnameWithTrailingSlash(uri.fsPath)
+
+            return sourceCodeUri.fsPath.startsWith(dirname)
+        })
+        .sort()
+        .reverse()
+
+    return parentProjectFiles.length === 0 ? undefined : parentProjectFiles[0]
+}
+
+export function isPublicClassSymbol(
+    document: Pick<vscode.TextDocument, 'getText'>,
+    symbol: vscode.DocumentSymbol,
+): boolean {
+    if (symbol.kind === vscode.SymbolKind.Class) {
+        const methodText: string = document.getText(symbol.range)
```

Comment:
What is `symbol.selectionRange` in this case? Is it just `MyClass`, or is it `public class MyClass`?

Reply:
Also, what if the user has non-compiling code like:

```csharp
private Foo {
    Console.WriteLine("publicclass")
```

Reply:
A Symbol wouldn't be emitted